sampler.go

```diff
@@ -449,11 +452,15 @@ func (s *RemotelyControlledSampler) IsSampled(id TraceID, operation string) (boo
 
 // Close implements Close() of Sampler.
 func (s *RemotelyControlledSampler) Close() {
-	s.RLock()
-	s.timer.Stop()
-	s.RUnlock()
+	if swapped := atomic.CompareAndSwapInt64(&s.closed, 0, 1); !swapped {
+		s.logger.Error("Repeated attempt to close the sampler is ignored")
+		return
+	}
 
-	s.pollStopped.Wait()
+	var wg sync.WaitGroup
+	wg.Add(1)
+	s.doneChan <- &wg
+	wg.Wait()
```

Comment:
Looks good in general, is there an advantage to waiting for the poller update to complete? (Wouldn't it be gc'd with simply the `return` on L493?

Reply:
it's a pattern we used in several places. The main objective is to ensure that the poller GR is about to exit and is guaranteed to not do any work, for example by receiving another ticket update and performing sampler update. We may not need this guarantee here, but I would leave it for symmetry with the remote reporter.

If such guarantee was not needed, a simpler pattern is to close the doneChan and not wait for the GR to report readiness to exit.

Reply:
Thanks for the explanation!

From my understanding of select case, only one case can be selected at a time.  Selecting the case with doneChan would result in the GR cleanly returning without doing any additional work and maintaining the guarantee, no?

Is symmetry at the cost of additional complexity important?

Reply:
`select` runs in a loop in a different GR from the one that calls Close(). Without a barrier Close may return prior to the loop GR exiting, so if any tests rely on having no updates after a call to Close we might break them.

I am not adding new complexity, the wait group was already in the implementation. I think my change makes that implementation simpler by localizing several vars that were previously spread out in different methods (like the ticket and the wait group itself). It also makes Close safe to call >1, something that we've seen people had trouble with in the reporter.